### PR TITLE
[mlir][Linalg] Allow more control in drop unit dims

### PR DESCRIPTION
**Pull request by sommerlukas**
Extend the ControlDropUnitDims struct to allow users of the `linalg::dropUnitDims` function more control over the behavior of the function.

The extended struct allows users to specify functions to control how the operands are collapsed and how the result is expanded to the original shape.

One example (and the motivation for this change) where this additional control is useful is to allow collapsing of tensors with an encoding, as demonstrated by the new test.

This is a breaking change. The new default behavior changes to abort the transformation if one of the operands cannot be collapsed or if the result cannot be expanded. This is the case for `memref` with non-identity layout and `tensor`s with an encoding.

### CODE REVIEW

**Review by rengolin — COMMENTED** [review]
Thanks! It's much clearer now.

I'll let others have time to review.

File: mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h [hunk 1]
```diff
@@ -516,6 +516,28 @@ LogicalResult vectorizeOpPrecondition(Operation *op,
...
 
+// Forward declaration
+struct ControlDropUnitDims;
+
+/// Collapse the given \p value to \p targetShape. The \p reassociation is used
+/// when `rankReductionStrategy` of \p control is set to
+/// `RankReductionStrategy::ReassociativeReshape`. Will return failure if the
+/// operand has memref type with a non-identity layout or tensor type with an
+/// encoding.
+FailureOr<Value> collapseValue(RewriterBase &rewriter, Location loc,
```
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h]: Hmm, on second thought, I think it might be better to make this method a static value in `ControlDropUnitDims` since this only is relevant for drop unit-dims. So something like

```
struct ControlDropUnitDims {

private:
  static CollapseFnTy defaultCollapseValueImpl;

  static ExpandFnTy defaultExpandValueImpl;
};
```

File: mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp [hunk 2]
```diff
@@ -264,23 +268,25 @@ expandValue(RewriterBase &rewriter, Location loc, Value result, Value origDest,
...
-    ArrayRef<int64_t> targetShape, ArrayRef<ReassociationIndices> reassociation,
-    ControlDropUnitDims::RankReductionStrategy rankReductionStrategy) {
+FailureOr<Value>
+linalg::collapseValue(RewriterBase &rewriter, Location loc, Value operand,
+                      ArrayRef<int64_t> targetShape,
+                      ArrayRef<ReassociationIndices> reassociation,
+                      const ControlDropUnitDims &control) {
   if (auto memrefType = dyn_cast<MemRefType>(operand.getType())) {
-    if (rankReductionStrategy ==
+    if (!memrefType.getLayout().isIdentity()) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/DropUnitDims.cpp]: I agree this check is needed, but it might be better to do this as a follow up. You actually dont need identity, you only need to check that the collapsed dimensions are contiguous. Since this isnt in the existing behavior, lets do this in a follow up.

**Review by MaheshRavishankar — COMMENTED** [review]
Looking good. Just a few minor comments, and should be good to go.

File: mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h (hunk 1 above)
Comment (sommerlukas) [reply mlir/include/mlir/Dialect/Linalg/Transforms/Transforms.h]: I've turned them into private `static` class functions. Using a `static` `std::function` member would require a global constructor, triggering a warning.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks! This looks good to me.

**Comment by banach-space** [comment]
> @rengolin @MaheshRavishankar @banach-space Are you OK if I merge this on Thursday (2025-12-18) or should we wait until after the holiday break due to the breaking nature of these changes?

To me this looks reasonable, but I don't use this logic  myself - at least not directly. @MaheshRavishankar and @hanhanW are experts in this area, so I trust their judgement. 

One note re testing ...

> With the changes from this PR, the transformation is now aborted in these cases, i.e., the operation is not changed.

When changing behaviour like this, I would expect to see some new negative tests that demonstrate cases in which the transformation bails out. Also, the two newly added tests exercise two edge cases:
* Encoding added to an operand.
* Encoding added to the output tensor.

What should happen when encoding is added to both? Why if different encodings are used? It feels like we are missing two more tests, no?
